### PR TITLE
Move impala op reg batch 1

### PR DESCRIPTION
**Pull request by matthewmturner**
@datapythonista @jreback 

I took a first stab at moving over some of the impala `_operation_registry` functions to `base_sql`, removing them from the `impala.compiler` (just commenting them out for now), and adding them to the `bigquery.compiler` registry.  

Looking at it now...im not sure this was the right way to do it.  Can you let me know your thoughts?

Is it ok to keep moving forward with this approach?

Or is it better to just use one operation registry at a time and import the functions from `base_sql` to `imapala.compiler` until i move the registry to `base_sql`?

Hope i didnt over complicate this.

Thx for your time and feedback.

### CODE REVIEW

File: ibis/impala/compiler.py [hunk 1]
```diff
@@ -1043,16 +1043,16 @@ def _sign(translator, expr):
 
 _operation_registry = {
```
Comment (datapythonista) [inline ibis/impala/compiler.py:1044]: You're removing those operations from Impala, so you need to add them back by extending the `operation_registry` you created in `base_sql`.

**Review by datapythonista — COMMENTED** [review]
The approach looks good.

Can you remove the functions you're moving instead of commenting them? Not much value on commenting since we're using git and we can restore them anytime easily.

I see you created like 10 functions in base_sql and you just removed four in Impala. I guess it's just because this is WIP, but I think this PR will be more reliable if you just move them, instead of copying them first, so we make sure nothing is left, and we don't need to keep counting how many functions we need to remove.

After you move something, if it's still used in Impala, you just import it from base_sql.

Thanks for working on this @matthewmturner, very useful, let me know when you got green CI and you're ready for a review/merge.

File: ibis/backends/bigquery/compiler.py [hunk 2]
```diff
@@ -338,7 +340,11 @@ def _formatter(translator, expr):
 }
 
 
-_operation_registry = impala_compiler._operation_registry.copy()
+# _operation_registry = impala_compiler._operation_registry.copy()
+_operation_registry = {
+    **impala_compiler._operation_registry.copy(),
+    **base_sql._operation_registry.copy(),
```
Comment (datapythonista) [inline ibis/backends/bigquery/compiler.py:346]: I'd assume that constructing a dictionary this way will make a copy, so the `.copy()` are probably not needed. If you can do a quick test to make sure that would be great.

Better to have `base_sql` first, and `impala` later (see comment below).

Also, you can delete the comment.

File: ibis/impala/compiler.py [hunk 3]
```diff
@@ -1162,7 +1007,8 @@ def _sign(translator, expr):
     ops.DayOfWeekName: _day_of_week_name,
 }
 
-_operation_registry.update(_binary_infix_ops)
+_operation_registry.update(binary_infix_ops)
+_operation_registry.update(base_sql._operation_registry)
```
Comment (datapythonista) [inline ibis/impala/compiler.py:1011]: It'd be better to do it the other way round. Start with `base_sql._operation_registry` (which you can make public removing the first underscore), and then updating with the Impala specific backend.

There is probably no overlap, but if something is defined in both, would be good to overwrite the base operation with the impala operation, not the other way round.

**Review by datapythonista — COMMENTED** [review]
Thanks @matthewmturner, really nice. Few comments, but looks great.
 
I think there are other backends that are also using the impala operator registry, not only bigquery (clickhouse, and I think some others too). You'll have to do the same you did for bigquery with all.

File: ibis/backends/bigquery/compiler.py (hunk 2 above)
Comment (matthewmturner) [reply ibis/backends/bigquery/compiler.py:346]: Sure.  I think you are right about the copy.  I did the following:

```
>>> a = {"A": 1, "B": 2}
>>> b = {**a}
>>> a is b
False
```

Is that a good test?

Comment (datapythonista) [reply ibis/backends/bigquery/compiler.py:346]: Yes, thanks, safe to remove the `.copy()`.

**Comment by matthewmturner** [comment]
> 
> 
> Thanks @matthewmturner, really nice. Few comments, but looks great.
> 
> I think there are other backends that are also using the impala operator registry, not only bigquery (clickhouse, and I think some others too). You'll have to do the same you did for bigquery with all.

I could be wrong but i dont think clickhouse uses impala, it looks like it just re-implemented the functions without importing.  The structure looks different there as well since it has a `operations.py` in addition to the `compiler.py`.

It looks like some functions may be implemented slightly differently as well.  For example, `fixed_arity`:

base_sql
```
def fixed_arity(func_name, arity):
    def formatter(translator, expr):
        op = expr.op()
        if arity != len(op.args):
            raise com.IbisError('incorrect number of args')
        return format_call(translator, func_name, *op.args)

    return formatter
```

clickhouse
```
def fixed_arity(func_name, arity):
    def formatter(translator, expr):
        op = expr.op()
        arg_count = len(op.args)
        if arity != arg_count:
            msg = 'Incorrect number of args {0} instead of {1}'
            raise com.UnsupportedOperationError(msg.format(arg_count, arity))
        return _call(translator, func_name, *op.args)

    return formatter
```

How should this be handled?

**Comment by datapythonista** [comment]
Yes, sorry, looks like clickhouse doesn't use Impala, it's bigquery, omnisci and spark for what I see. The CI should hopefully tell anyway.

Don't worry about the code already duplicated for now. Feel free to create an issue to take care of avoiding that duplication of functions later. But I think avoiding weird dependencies among backends is enough work for now, and will be extremely valuable.

**Comment by jreback** [comment]
@matthewmturner yeah this looks like a good approach. if you can merge master and hopefully this will be green. we can merge, then ideally you can adjust the bigquery, spark & omnisci to use this (agree clickhoouse is out of scope for now, though certainly happy to have that at some point to use the same).

File: ibis/backends/bigquery/compiler.py [hunk 4]
```diff
@@ -338,7 +340,11 @@ def _formatter(translator, expr):
 }
 
 
-_operation_registry = impala_compiler._operation_registry.copy()
+# _operation_registry = impala_compiler._operation_registry.copy()
```
Comment (jreback) [inline ibis/backends/bigquery/compiler.py:343]: can be in a followon; if we are importing from / to another module then these should be non-private (no leading _)

**Review by jreback — APPROVED** [review]

**Comment by matthewmturner** [comment]
@jreback sure - merged master.  I also already made updates to bigquery, spark, & omnisci to use base_sql in addition to impala (once everything moved from impala of course that will be removed).   was there another change you were expecting to those backends?

File: ibis/backends/bigquery/compiler.py (hunk 4 above)
Comment (matthewmturner) [reply ibis/backends/bigquery/compiler.py:343]: noted. i had left the original impala updates as is. but will update these on next batch.

**Comment by jreback** [comment]
> @jreback sure - merged master. I also already made updates to bigquery, spark, & omnisci to use base_sql in addition to impala (once everything moved from impala of course that will be removed). was there another change you were expecting to those backends?

nope that's good.

**Comment by matthewmturner** [comment]
> 
> 
> > @jreback sure - merged master. I also already made updates to bigquery, spark, & omnisci to use base_sql in addition to impala (once everything moved from impala of course that will be removed). was there another change you were expecting to those backends?
> 
> nope that's good. goal of course is to move out as much common code as possible and centrailze w/o the heavy deps as existed before.

sounds good. i also just fixed for the omnisci CI test failure.  isort was still throwing errors though in CI.  will see how it looks after these tests.

**Comment by jreback** [comment]
@matthewmturner ping when this is green

**Comment by matthewmturner** [comment]
> 
> 
> @matthewmturner ping when this is green

still having issues in CI with isort / asv missing in isort section of `setup.cfg` again.  it looks like CI is using isort 5.6.4 and im on 5.5.3 locally, will update that and try again.

**Comment by matthewmturner** [comment]
@jreback i think were good. let me know if you need anything else updated.

**Comment by jreback** [comment]
thanks @matthewmturner

**Pull request merged**